File: Sudoku Solver_hard.cpp
class Solution {
public:
    pair<int,int> Missingindex(vector<vector<char>>& matrix){
        int i, j, m = matrix.size(), n = matrix[0].size();
        
        for(i = 0; i < m; i++){
            for(j = 0; j < n; j++)
                if (matrix[i][j] == '.')
                    return {i,j};
        }
        
        return {-1,-1};
    }
    bool Valid(vector<vector<char>>& matrix, int num, pair<int,int> mv){
        int i, j, m = matrix.size(), n = matrix[0].size(), r, c;
        
        r = mv.first;
        c = mv.second;
        for(i = 0; i < n; i++)
            if (matrix[r][i] == num + '0')
                return false;
        for(i = 0; i < m; i++)
            if (matrix[i][c] == num + '0')
                return false;
        m = 3 * (r / 3);
        n = 3 * (c / 3);
        for(i = m; i < m + 3; i++){
            for(j = n; j < n + 3; j++)
                if (matrix[i][j] == num + '0')
                    return false;
        }
        return true;
    }
    bool SolveSudoku(vector<vector<char>>& board){
        int i;
        pair<int,int> mv = {};
        
        mv = Missingindex(board);
        if (mv.first == -1)
            return true;
        
        for(i = 1; i <= 9; i++){
            if (Valid(board, i, mv)){
                board[mv.first][mv.second] = i + '0';
                if(!SolveSudoku(board))
                    board[mv.first][mv.second] = '.';
                else
                    return true;
            }
        }
        return false;
    }
    void solveSudoku(vector<vector<char>>& board) {
        SolveSudoku(board);
        return;
    }
};

File: Sudoku Solver_hard.py
class Solution:
    def Missingindex(self, matrix: List[List[str]]) -> [int, int]:
        m, n = len(matrix), len(matrix[0])

        for i in range(m):
            for j in range(n):
                if matrix[i][j] == '.':
                    return [i, j]
        return [-1, -1]

    def Valid(self, num: int, x: int, y: int, matrix: List[List[str]]) -> bool:
        m, n = len(matrix), len(matrix[0])

        for i in range(m):
            if matrix[i][y] == str(num):
                return False
        for i in range(n):
            if matrix[x][i] == str(num):
                return False
        m, n = 3 * (x // 3), 3 * (y // 3)
        for i in range(m, m + 3):
            for j in range(n, n + 3):
                if matrix[i][j] == str(num):
                    return False
        return True

    def SolveSudoku(self, matrix: List[List[str]]) -> bool:
        x, y = self.Missingindex(matrix)
        if x == -1:
            return True

        for i in range(1, 10):
            if self.Valid(i, x, y, matrix):
                matrix[x][y] = str(i)
                if self.SolveSudoku(matrix) == False:
                    matrix[x][y] = '.'
                else:
                    return True
        return False

    def solveSudoku(self, board: List[List[str]]) -> None:
        """
        Do not return anything, modify board in-place instead.
        """
        self.SolveSudoku(board)